android-core/src/main/java/com/mparticle/internal/DeviceAttributes.kt

```diff
@@ -0,0 +1,343 @@
+                    editor.putLong(PrefKeys.UPGRADE_DATE, upgradeDate)
+                }
+                countSinceUpgrade += 1
+                userStorage.launchesSinceUpgrade = countSinceUpgrade
+
+                attributes.put(MessageKey.LAUNCH_COUNT_SINCE_UPGRADE, countSinceUpgrade)
+                attributes.put(MessageKey.UPGRADE_DATE, upgradeDate)
+            } catch (e: PackageManager.NameNotFoundException) {
+                // ignore missing data
+            }
+            val instance = MParticle.getInstance()
+            if (instance != null) {
+                attributes.put(MessageKey.ENVIRONMENT, ConfigManager.getEnvironment().value)
+            }
+            attributes.put(MessageKey.INSTALL_REFERRER, preferences.getString(PrefKeys.INSTALL_REFERRER, null))
+
+            val install = preferences.getBoolean(PrefKeys.FIRST_RUN_INSTALL, true)
+            attributes.put(MessageKey.FIRST_SEEN_INSTALL, install)
+            editor.putBoolean(PrefKeys.FIRST_RUN_INSTALL, false)
+        } catch (e: Exception) {
+            // again different devices can do terrible things, make sure that we don't bail out completely
+            // and return at least what we've built so far.
+        } finally {
+            editor.apply()
+        }
+        return attributes
+    }
+
+    fun updateInstallReferrer(context: Context, attributes: JSONObject) {
+        val preferences = context.getSharedPreferences(Constants.PREFS_FILE, Context.MODE_PRIVATE)
+        try {
+            attributes.put(MessageKey.INSTALL_REFERRER, preferences.getString(PrefKeys.INSTALL_REFERRER, null))
+        } catch (ignored: JSONException) {
+            // this, hopefully, should never fail
+        }
+    }
+
+    /**
+     * Generates a collection of device attributes that will not change during an app's process.
+     *
+     *
+     * This contains logic that MUST only be called once per app run.
+     *
+     * @param appContext the application context
+     * @return a JSONObject of device-specific attributes
+     */
+    fun getStaticDeviceInfo(appContext: Context): JSONObject {
+        val attributes = JSONObject()
+
+        try {
+            // device/OS attributes
+            attributes.put(MessageKey.BUILD_ID, Build.ID)
+            attributes.put(MessageKey.BRAND, Build.BRAND)
+            attributes.put(MessageKey.PRODUCT, Build.PRODUCT)
+            attributes.put(MessageKey.DEVICE, Build.DEVICE)
+            attributes.put(MessageKey.MANUFACTURER, Build.MANUFACTURER)
+            attributes.put(MessageKey.PLATFORM, operatingSystemString)
+            attributes.put(MessageKey.OS_VERSION, Build.VERSION.SDK)
+            attributes.put(MessageKey.OS_VERSION_INT, Build.VERSION.SDK_INT)
+            attributes.put(MessageKey.MODEL, Build.MODEL)
+            attributes.put(MessageKey.RELEASE_VERSION, Build.VERSION.RELEASE)
+
+            val application = appContext as Application
```

Comment:
Are we sure that `appContext` is always an `Application`? Can we add a check for that here and handle the case where it isn't? Or maybe better, if it's always an `Application` can we just change the input parameter to `(applicaiton: Application)` to ensure it?